### PR TITLE
CMake migration rule, helper function and LIT test for target_link_libraries

### PR DESCRIPTION
**Pull request by daiyaan-ahmed6**
Adds CMake migration rule, helper function and LIT test for target_link_libraries

Signed-off-by: Daiyaan Ahmed <daiyaan.ahmed@intel.com>

### CODE REVIEW

File: clang/test/dpct/cmake_migration/case_004/expected.txt [hunk 1]
```diff
@@ -0,0 +1,18 @@
+target_link_sycl_libraries(main PRIVATE cxxopts fmt::fmt -lcublas)
+
+target_link_sycl_libraries(transformer_engine PUBLIC 
+   CUDA::cublas
+   CUDA::cuda_driver
+   CUDA::cudart
+   CUDA::nvrtc
+   CUDA::nvToolsExt
+   CUDNN::cudnn)
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migration/case_004/expected.txt]: After a further thought, target_link_libraries() is not a cuda specific cmake syntax, so  target_link_sycl_libraries is not needed introduced here, we can map cuda libraries to sycl mappings if the mapping is available, or "" if mapping is not available.

Comment (daiyaan-ahmed6) [reply clang/test/dpct/cmake_migration/case_004/expected.txt]: Some libraries need to be linked based on windows specific or linux specific names, for this reason we had discussed using a helper function. Do you think it is not required anymore?

Comment (tomflinda) [reply clang/test/dpct/cmake_migration/case_004/expected.txt]: @daiyaan-ahmed6  synced with @zhimingwang36:
target_link_libraries() is not a cuda specific cmake syntax, so target_link_sycl_libraries is not needed introduced here.
#1. We map cuda library to sycl mapping if the mapping is available, or "" if mapping is not available.
#2. Some libraries need to be linked based on windows specific or linux specific names, we import a variable in helper function. Take "CUDNN::cudnn" for example, we introduce variable "dnnlib", and set its value to "dnnl.lib" on Windows, and to "dnnl" on Linux, then migrate target_link_sycl_libraries(transformer_engine PUBLIC CUDNN::cudnn) to target_link_sycl_libraries(transformer_engine PUBLIC dnnlib).

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 2]
```diff
@@ -264,3 +264,154 @@
...
+- Rule: rule_target_link_libraries_CUDA_LIBRARIES
+  Kind: CMakeRule
+  Priority: Fallback
+  CmakeSyntax: CUDA_LIBRARIES_target_link_libraries
+  In: target_link_libraries(${CULibraries})
+  Out: target_link_libraries(${CULibraries})
+  Subrules:
+    CULibraries:
+      In: |
+        \${CUDA_LIBRARIES}
```
Comment (tomflinda) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml:369]: Pls add comments why ${CUDA_LIBRARIES} should be removed.

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 3]
```diff
@@ -264,3 +264,154 @@
...
+
+- Rule: rule_target_link_libraries_libstdc++
+  Kind: CMakeRule
+  Priority: Fallback
+  CmakeSyntax: libstdc++_target_link_libraries
+  In: target_link_libraries(${StaticLibraries})
+  Out: target_link_libraries(${StaticLibraries})
+  Subrules:
+    StaticLibraries:
+      In: -static-libstdc++
```
Comment (tomflinda) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml:507]: Pls add comments why -static-libstdc++ should be removed.

File: clang/tools/dpct/cmake/dpct.cmake [hunk 4]
```diff
@@ -78,3 +78,9 @@ endmacro()
 # Always set SYCL_HAS_FP16 to true to assume SYCL device to support float16
 message("dpct.cmake: SYCL_HAS_FP16 is set true by default.")
 set(SYCL_HAS_FP16 TRUE)
+
+if(WIN32)
```
Comment (tomflinda) [inline clang/tools/dpct/cmake/dpct.cmake:94]: Pls submit behavior test for DNN_LIB in SYCLomatic-test, and sycl version test in https://github.com/intel-sandbox/cmake_script_migration/tree/master/lit/target_link_libraries/ to verify DNN_LIB

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml (hunk 3 above)
Comment (daiyaan-ahmed6) [reply clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml:507]: Added. Done

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml (hunk 2 above)
Comment (daiyaan-ahmed6) [reply clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml:369]: Added. Done.

File: clang/test/dpct/cmake_migration/case_004/expected.txt (hunk 1 above)
Comment (daiyaan-ahmed6) [reply clang/test/dpct/cmake_migration/case_004/expected.txt]: Done.

File: clang/tools/dpct/cmake/dpct.cmake (hunk 4 above)
Comment (daiyaan-ahmed6) [reply clang/tools/dpct/cmake/dpct.cmake:94]: Behavior test for target_link_libs: https://github.com/oneapi-src/SYCLomatic-test/pull/577

**Review by zhiweij1 — APPROVED** [review]

**Pull request merged**